### ISSUE TITLE
Make Jax work better with threading?

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#17457]
Jax seems to hang when Orion (a hyperparameter optimizer) calls it in multi-threaded mode: https://github.com/Epistimio/orion/issues/1117.  Removing the line `a = jnp.zeros(10)` makes this pass.  This suggests that Jax is doing some kind of initialization that doesn't carry over to the other threads that Orion creates.  Can we make this work somehow?

I don't know what exactly is going wrong, but it would be beautiful form a user perspective if all I had to do would be to create a context manager like this:

```python
from typing import Any

import jax
import jax.numpy as jnp
from jax.random import PRNGKey, truncated_normal
from orion.client import build_experiment

a = jnp.zeros(10)

def experiment_f(x: float) -> list[dict[str, Any]]:
    # This function is called once for each thread.
    with jax.Session(enable_x64=True,
                     enable_custom_prng=True,
                     compilation_cache='./cache'):
        key = PRNGKey(123)
        print("DIE")  # This is printed thrice
        truncated_normal(key, -2.0, 2.0, (85, 85))
        print("DEAD")   # This is never printed
        return [{'name': 'loss', 'type': 'objective', 'value': 1.0}]

workers = 3
space = {'x': "loguniform(1e-4, 1e2)"}
experiment = build_experiment(name='blah', space=space,
                              max_trials=10, max_broken=1,
                              algorithm={"tpe": {"n_initial_points": 5}})
with experiment.tmp_executor("joblib", n_workers=workers):
    experiment.workon(experiment_f, workers)
```

Also, would the various threads share the in-memory compilation cache?  What about the on-disk compilation cache (`jax.experimental.compilation_cache`)?

I also noticed that this seems to call `_initialize_jax_jit_thread_local_state` only once.

**Comment by NeilGirdhar** [comment]
I also noticed that this seems to call `_initialize_jax_jit_thread_local_state` only once.

**Comment by NeilGirdhar** [comment]
Closing as duplicate of these related issues:

https://github.com/google/jax/issues/3691
https://github.com/google/jax/issues/4228
